tutorial/kalman/tutorial-ukf-nonlinear-complex-example.cpp

```diff
@@ -0,0 +1,633 @@
+   along the x and y axis, the measurement vector can be written as:
+   \f{eqnarray*}{
+        \textbf{z}[2i] &=& \sqrt{(p_x^i - x)^2 + (p_y^i - y)^2} \\
+        \textbf{z}[2i+1] &=& \tan^{-1}{\frac{p_y^i - y}{p_x^i - x}} - \theta
+   \f}
+
+ * Some noise is added to the measurement vector to simulate measurements which are
+ * not perfect.
+
+   The mean of several angles must be computed in the Unscented Transform. The definition we chose to use
+   is the following:
+
+   \f$ mean(\boldsymbol{\theta}) = atan2 (\frac{\sum_{i=1}^n \sin{\theta_i}}{n}, \frac{\sum_{i=1}^n \cos{\theta_i}}{n})  \f$
+
+   As the Unscented Transform uses a weighted mean, the actual implementation of the weighted mean
+   of several angles is the following:
+
+   \f$ mean_{weighted}(\boldsymbol{\theta}) = atan2 (\sum_{i=1}^n w_m^i \sin{\theta_i}, \sum_{i=1}^n w_m^i \cos{\theta_i})  \f$
+
+   where \f$ w_m^i \f$ is the weight associated to the \f$ i^{th} \f$ measurements for the weighted mean.
+
+   Additionnally, the addition and substraction of angles must be carefully done, as the result
+   must stay in the interval \f$[- \pi ; \pi ]\f$ or \f$[0 ; 2 \pi ]\f$ . We decided to use
+   the interval \f$[- \pi ; \pi ]\f$ .
+*/
+
+// UKF includes
+#include <visp3/core/vpUKSigmaDrawerMerwe.h>
+#include <visp3/core/vpUnscentedKalman.h>
+
+// ViSP includes
+#include <visp3/core/vpConfig.h>
+#include <visp3/core/vpGaussRand.h>
+#ifdef VISP_HAVE_DISPLAY
+#include <visp3/gui/vpPlot.h>
+#endif
+
+#if (VISP_CXX_STANDARD >= VISP_CXX_STANDARD_11)
+namespace
+{
+/**
+ * \brief Normalize the \b angle in the interval [-Pi; Pi].
+ *
+ * \param[in] angle Angle to normalize.
+ * \return double Normalized angle.
+ */
+double normalizeAngle(const double &angle)
+{
+  double angleIn0to2pi = vpMath::modulo(angle, 2. * M_PI);
+  double angleInMinPiPi = angleIn0to2pi;
+  if (angleInMinPiPi > M_PI) {
+    // Substract 2 PI to be in interval [-Pi; Pi]
+    angleInMinPiPi -= 2. * M_PI;
+  }
+  return angleInMinPiPi;
+}
+
+/**
+ * \brief Compute the weighted mean of measurement vectors.
+ *
+ * \param[in] measurements The measurement vectors.
+ * \param[in] wm The associated weights.
+ * \return vpColVector
+ */
+vpColVector measurementMean(const std::vector<vpColVector> &measurements, const std::vector<double> &wm)
+{
+  const unsigned int nbPoints = measurements.size();
+  const unsigned int sizeMeasurement = measurements[0].size();
+  const unsigned int nbLandmarks = sizeMeasurement / 2;
+  vpColVector mean(sizeMeasurement, 0.);
+  std::vector<double> sumCos(nbLandmarks, 0.);
+  std::vector<double> sumSin(nbLandmarks, 0.);
+  for (unsigned int i = 0; i < nbPoints; ++i) {
+    for (unsigned int j = 0; j < nbLandmarks; ++j) {
+      mean[2*j] += wm[i] * measurements[i][2*j];
+      sumCos[j] += wm[i] * std::cos(measurements[i][2*j+1]);
+      sumSin[j] += wm[i] * std::sin(measurements[i][2*j+1]);
+    }
```

Comment:
With respect to sonaqube quality rules add parenthesis
```
      sumCos[j] += wm[i] * std::cos((measurements[i][2*j) + 1]);
      sumSin[j] += wm[i] * std::sin((measurements[i][2*j) + 1]);
```

Reply:
Done